create dynamic memory management concept (#773)
exercism/c: concepts/dynamic-memory-management/about.md

A detailed analysis of variable length arrays is outside the scope of this article.

## malloc

`malloc` is a function declared in `stdlib.h` which is used for allocating dynamic memory.
If the allocation is successful, a pointer is returned which points to the address of the first byte of the allocated memory.
If the allocation is not successful, `NULL` is returned.
It is prudent to check the result of `malloc` for `NULL` and gracefully handle it if it fails.
The pointer returned is of type `pointer-to-void` which can roughly be considered as a "generic" pointer that can be converted to the specific type used by the memory.
`malloc` has one parameter, which is the size of the desired memory in bytes.
The total size for allocating an array is the size of each element's data type multiplied by the number of elements.
The `sizeof` operator is commonly used to get the size of each element's data type.
The following example demonstrates this

```c
#include <stdio.h>
#include <stdlib.h>

int * my_function(int count) {
    // pointer-to-void from malloc is converted to a pointer-to-int

>     int * numbers = malloc(sizeof(int) * count);

Author: Thanks, I think I want to leave it as is to be a bit more straightforward.

The pointer type was implicitly converted from pointer-to-void to pointer-to-int.
It could also have been explicitly converted like so : `int * numbers = (int *) malloc(sizeof(int) * count);`.
Whether the pointer-to-int points to a single `int` or to an array of `ints` is something the programmer can deal with as seen fit.
For example

```c
#include <stdio.h>
#include <stdlib.h>

int * my_function(int count) {
    int * numbers = malloc(sizeof(int) * count);
    return numbers;
}

int main() {
    int * numbers = my_function(1);
    if (!numbers) return -1;
    // treat numbers as a single value
    *numbers = 1;
    // prints numbers is 1
    printf("numbers is %d\n", *numbers);
    // treat numbers as an array
    // prints The first element of the array is 1
    printf("The first element of the array is %d\n", numbers[0]);
    numbers[0] = 2;
    // prints The first element of the array is 2
    printf("The first element of the array is %d\n", numbers[0]);
    // treat numbers as a single value again
    // prints numbers is 2
    printf("numbers is %d\n", *numbers);    
    free(numbers);
    numbers = NULL;    
}

> ```
> If the value(s) will definitely be written to before being read, then to use `malloc` can be an efficient way to allocate memory.

Reviewer: ````suggestion
```

If the value(s) will definitely be written to before being read, then to use `malloc` can be an efficient way to allocate memory.
````

But, if there is a chance the value(s) will be read before being intialized, then to use `malloc` is risky.
A safer way to allocate memory is to use `calloc`, which initializes the newly allocated memory to zero.

Note in the examples above that `free(numbers)` was used.

## free

`free` is a function declared in `stdlib.h` which is used for deallocating memory which has been dynamically allocated.
Dynamically allocated memory stays claimed in memory until released by `free`.
Some operating systems may automatically free dynamically allocated memory when the program finishes, but it is best practice not to depend on that.
`free` has one parameter, which is a pointer to the memory which was dynamically allocated.
The use or lack of use of `free` is associated with several memory issues.

#### memory leak

> Not freeing memory after it is no longer needed is referred to as a memory leak.

Reviewer: ```suggestion
Not freeing dynamically allocated memory after the reference to that memory reaches the end of its lifetime, or is otherwise destroyed, is referred to as a memory leak.
```

    // The first element of the array is 18178656
    printf("The first element of the array is %d\n", numbers[0]);
}
```

The `numbers` pointer, after it has been freed, is referred to as a dangling pointer.
A dangling pointer can refer to other memory than that for which it was intended.
Thus, a dangling pointer can lead to data corruption and arbitrary code substitution and execution.
A pointer to allocated memory which has not been cleared by setting to `NULL` can be used to hack the program.
In the above example, if `numbers` was set to `NULL` like so: `numbers = NULL;` after being freed, the following `printf` would result in program termination, perhaps reporting a segmentation fault when attempting to read `numbers[0]`.

#### double free

If `free` is called twice on a pointer to dynamically allocated memory, it results in undefined behavior (**UB**.)
This means the compiler can arbitrarily handle it any way it wants to.
The following example, compiled with two different compilers, can have two different outcomes

```c
#include <stdio.h>
#include <stdlib.h>

int * my_function(int count) {
    // pointer-to-void from malloc is converted to a pointer-to-int

>     int * numbers = (int *) malloc(sizeof(int) * count);

Reviewer: ```suggestion
    int * numbers = malloc(sizeof(int) * count);
```

    return numbers;
}

int main() {
    int * numbers = my_function(1);
    if (!numbers) return -1;
    *numbers = 1;
    // prints numbers is 1
    printf("numbers is %d\n", *numbers);
    free(numbers);
    // first compiler generates runtime error: double free detected in tcache 2
    // second compiler generates no error
    free(numbers);
}
```

> The first compiler generated no runtime error when `numbers` was set to `NULL` between each call to `free`, but that would not be guaranteed to work with a third compiler.

Reviewer: I read this as there should be a `numbers = NULL;` line somewhere in the above example, specifically between the two calls to `free()`. But I don't see such. 
Do I misunderstand?

Author: Good catch! I just put it in a commented line, since the program was run both with and without setting `numbers = NULL`.

}

int * my_function_2 (int * ptr , int count) {
    // reallocates the memory to the size of an int multiplied by count
    int * numbers = realloc(ptr, sizeof(int) * count);
    // if realloc fails, exit the program
    if (!numbers) exit(-1);
    return numbers;
}

int main() {
    int * numbers = my_function(3);
    if (!numbers) return -1;
    numbers[0] = 1;
    numbers[1] = 2;
    numbers[2] = 3;
    // expand the size of the array
    numbers = my_function_2(numbers, 6);
    // prints The first element of the array is 1
    printf("The first element of the array is %d\n", numbers[0]);
    // prints The sixth element of the array is [whatever is there]
    // the additional memory is not initialized, so it could be
    // 0 or it could be 18728 or anything else
    printf("The sixth element of the array is %d\n", numbers[5]);
    numbers[5] = 6;
    // shrink the size of the array
    numbers = my_function_2(numbers, 3);
    // prints The sixth element of the array is [whatever is there]
    // the removed memory is not initialized, so it could be
    // 6 or it could be 43164 or anything else
    printf("The sixth element of the array is %d\n", numbers[5]);

> }

Reviewer: Should there be a `free()` somewhere in this example?

Author: Yes, there absolutely should. Good catch! Thanks!
